File: Elixir/conway.exs
defmodule Conway do
    defstruct [:height, :width, :rows] 

    def grid(h, w) do
        for _ <- 0..h-1 do
            for _ <- 0..w-1, do: 0
        end
    end

    def new(h, w) do
        %__MODULE__{
            height: h,
            width: w,
            rows: grid(h, w),
        }
    end

    def draw(row) do
        Enum.map(row, fn e ->
            case e do
                0 -> "."
                1 -> "#"
            end
        end)
    end

    def print(g) do
        Enum.map(g.rows, fn row ->
            IO.puts(Enum.join(draw(row), ""))
        end)
    end

    def assign(g, y, x, state) do
        row = Enum.at(g.rows, y)
        updated_row = List.replace_at(row, x, state)
        rows = List.replace_at(g.rows, y, updated_row)
        %{g | rows: rows}
    end

    def mod(x, y) when x > 0, do: rem(x, y)
    def mod(x, y) when x < 0, do: rem(y + x, y)
    def mod(0, _), do: 0

    def query(g, y, x) do
        row = Enum.at(g.rows, mod(y, g.height))
        Enum.at(row, mod(x, g.width))
    end

    def count_neighbors(g, y, x) do
        n  = query(g, y + 1, x    )
        ne = query(g, y + 1, x + 1)
        e  = query(g, y,     x + 1)
        se = query(g, y - 1, x + 1)
        s  = query(g, y - 1, x    )
        sw = query(g, y - 1, x - 1)
        w  = query(g, y,     x - 1)
        nw = query(g, y + 1, x - 1)
        n + ne + e + se + s + sw + w + nw
    end

    def game_logic(1, neighbors) do
        if neighbors < 2 or neighbors > 3 do
            0
        else
            1
        end
    end
    def game_logic(0, neighbors) do
        if neighbors == 3 do
            1
        else
            0
        end
    end

    def step_cell(g, y, x) do
        state = query(g, y, x)
        neighbors = count_neighbors(g, y, x)
        game_logic(state, neighbors)
    end

    def simulate(g) do
        res = for y <- 0..g.height-1 do
            for x <- 0..g.width-1, do: step_cell(g, y, x)
        end
        res
    end

    def clear_screen do
        IO.puts("\ec")
    end

    def demo_grid do
        height = 10
        width = 20
        g = new(height, width)
        g = assign(g, 0, 3, 1)
        g = assign(g, 1, 4, 1)
        g = assign(g, 2, 2, 1)
        g = assign(g, 2, 3, 1)
        assign(g, 2, 4, 1)
    end

    def run_simulation(g) do
        clear_screen()
        nrows = simulate(g)
        g = %{g | rows: nrows}
        print(g)
        Process.sleep(100)
        run_simulation(g)
    end

    def start do
        g = demo_grid()
        run_simulation(g)
    end
end

Conway.start

# elixir conway.exs

File: Swift/conway.swift
class Conway {
    let width: Int
    let height: Int
    var grid: Array<Array<Int>>

    init(width: Int, height: Int) {
        self.width = width
        self.height = height
        let row = Array(repeating: 0, count: width)
        self.grid = Array(repeating: row, count: height)
    }

    func draw(value: Int) -> Character {
        if value == 1 {
            return "#"
        }
        return "."
    } 

    func printGrid() {
        for row in self.grid {
            for e in row {
                print(draw(value: e), terminator: "")
            }
            print()
        }
    }

    func mod(x: Int, y: Int) -> Int {
        if x > 0 {
            return x % y
        }
        if x < 0 {
            return (y + x) % y
        }
        return 0
    }

    func query(x: Int, y: Int) -> Int {
        return self.grid[mod(x: x, y: self.height)][mod(x: y, y: self.width)]
    }

    func assign(x: Int, y: Int, state: Int) {
        self.grid[mod(x: x, y: self.height)][mod(x: y, y: self.width)] = state
    }

    func countNeighbors(x: Int, y: Int) -> Int {
        let n  = query(x: x + 1, y: y    )
        let ne = query(x: x + 1, y: y + 1)
        let e  = query(x: x,     y: y + 1)
        let se = query(x: x - 1, y: y + 1)
        let s  = query(x: x - 1, y: y    )
        let sw = query(x: x - 1, y: y - 1)
        let w  = query(x: x,     y: y - 1)
        let nw = query(x: x + 1, y: y - 1)
        return n + ne + e + se + s + sw + w + nw
    }

    func gameLogic(state: Int, neighbors: Int) -> Int {
        if state == 1 {
            if neighbors < 2 {
                return 0
            }
            if neighbors > 3 {
                return 0
            }
            return state
        } else {
            if neighbors == 3 {
                return 1
            }
            return state
        }
    }

    func stepCell(x: Int, y: Int) -> Int {
        let state = query(x: x, y: y)
        let neighbors = countNeighbors(x: x, y: y)
        let nextState = gameLogic(state: state, neighbors: neighbors)
        return nextState
    }

    func simulate() {
        let row = Array(repeating: 0, count: self.width)
        var newGrid: Array<Array<Int>> = Array(repeating: row, count: self.height)
        
        for x in 0..<self.height {
            for y in 0..<self.width {
                newGrid[x][y] = stepCell(x: x, y: y)
            }
        }
        self.grid = newGrid
    }

    func clearScreen() {
        print("\u{001B}[2J");
    }
}

var c: Conway = Conway(width: 20, height: 10)

c.assign(x: 0, y: 3, state: 1)
c.assign(x: 1, y: 4, state: 1)
c.assign(x: 2, y: 2, state: 1)
c.assign(x: 2, y: 3, state: 1)
c.assign(x: 2, y: 4, state: 1)

c.printGrid()

while true {
    c.clearScreen()
    c.simulate()
    c.printGrid()
    try await Task.sleep(nanoseconds: 100 * 1000000)
}

// swift conway.swift
